README.md
EClet
=====

[![Build Status](https://travis-ci.org/cryptotronix/EClet.png)](https://travis-ci.org/cryptotronix/EClet)

Driver for the Cryptotronix EClet containing an Atmel ATECC108

Status
---

This software is in ***ALPHA***. I have tested the below commands, but some there are some features and documentation that I would like to finish. If you use this software, it will configure your ECC108 in a non-reversible way. It will allow you to sign and verify with P256 keys but future features may be incompatible.

You can following along with the release status [here](https://github.com/cryptotronix/EClet/issues?milestone=1&state=open).

Building
----

This project uses Autotools.

It requires [libcrypti2c](https://github.com/cryptotronix/libcrypti2c), also in ***ALPHA***. If you use the [autogen](https://github.com/cryptotronix/EClet/blob/master/autogen.sh) script, it will automatically build and install the library.

Hardware
---

Product page will be available shortly. This software will also work on the [CryptoCape](http://www.cryptotronix.com/products/cryptocape).

Running
---

see `./eclet --help` for full details.  The default I2C bus is `/dev/i2c-1` and this can be changed with the `-b` option.

Root
---

You'll need to run as root to access `/dev/i2c*` initially.  You can change this by adding your user to the `i2c` group with:

`sudo usermod -aG i2c user`

Or:

`sudo chmod o+rw /dev/i2c*`


Currently supported commands:

### state
```bash
eclet state
Factory
```

This is the first command you should run and verify it's in the Factory state.  This provides the assurance that the device has not been tampered during transit.

### personalize
```bash
eclet personalize
```

This is the second command you should run.  On success it will not output anything. It configures all slots (0-16) to be holders for P-256 ECC private keys, except slot 8, which is reserved for future use. Keys are not generated at this time. Each key must be individually generated with the `gen-key` command.

### random
```bash
eclet random
62F95589AC76855A8F9204C9C6B8B85F06E6477D17C3888266AEE8E1CBD65319
```
### serial-num
```bash
eclet serial-num
0123XXXXXXXXXXXXEE
```
X's indicate the unique serial number.

### gen-key
```bash
eclet gen-key
04EED1CB629CF87F8BF6419986F990B92EA3DFA14CDAF70EB3E8DA8F9C9504DBC5B040D6480E88F895E9E1D4477970329B060450C80E1816EFED7B0FA49868CAEB
```

The device will internally create an P-256 ECC key and return the public key. The format of the public key is 0x04 + X + Y. Specify which slot to create a key (0-7, 9-15) with the `-k` option. Currently running this command multiple times will overwrite the public key, see this [issue](https://github.com/cryptotronix/EClet/issues/1).

### sign
```bash
eclet sign -f ChangeLog
3BAEB5705D8765B34B389F1768BAC783FCA786AB64A760D10DD133C86E5892A7A790E424C8E1540551C99FBE4F9F531B504A6004F08F3E0D4E42E96BBDE5C179
```

Performs an ECDSA signature. Data can be specified as a file with the `-f` option or passed via `stdin`. The data will be SHA256 hashed prior to signing. The result is the signature in the format: R + S.

### verify
```bash
eclet verify -f ChangeLog --signature C650D1A30194AD68F60F40C321FB084F6177BEDAC74D0F0C276ED35B00249AC8CF3E96FB7AB14AA48223FBA2E5DD9BCAE232BF963755C42F8FD9BD77FC145D41 --public-key 049B4A517704E16F3C99C6973E29F882EAF840DCD125C725C9552148A74349EB77BECB37AA2DB8056BAF0E236F6DCFEC2C5A9A0F23CEFD8A9DC1F4693718E725D2
```

Verifies an ECDSA signature using the device. You specify the data (which will be SHA256 hashed), the signature (R+S), and the public key (0x04+X+Y). Returns a `0` exit code on success.

### offline-verify-sign
```bash
eclet offline-verify-sign -f ChangeLog --signature C650D1A30194AD68F60F40C321FB084F6177BEDAC74D0F0C276ED35B00249AC8CF3E96FB7AB14AA48223FBA2E5DD9BCAE232BF963755C42F8FD9BD77FC145D41 --public-key 049B4A517704E16F3C99C6973E29F882EAF840DCD125C725C9552148A74349EB77BECB37AA2DB8056BAF0E236F6DCFEC2C5A9A0F23CEFD8A9DC1F4693718E725D2
```

Same as `verify` except it *does not* use the device and can be run on a system with one. It uses the software ECDSA implementation provided by `libcrypti2c`.

Options
---

Options are listed in the `--help` command, but a useful one, if there are issues, is the `-v` option.  This will dump all the data that travels across the I2C bus with the device.

Support
---

IRC: Join the `#cryptotronix` channel on freenode.

Mailing lists: `hashlet-announce` and `hashlet-users` are open for subscriptions [here](https://savannah.nongnu.org/mail/?group=hashlet).

Contributing
---
See the wiki page on [contributing](https://github.com/cryptotronix/hashlet/wiki/Contributing).


![GPLv3](https://www.gnu.org/graphics/gplv3-127x51.png)


NEWS
Eclet NEWS -- history of user-visible changes

Copyright (C) 2014 Cryptotronix, LLC.

* EClet 0.1.0 - Beta Release - May 2014

The initial release of EClet is concerned a Beta release.

** Features

*** Command line interface

The following commands are supported:

**** personalize
**** random
**** verify
**** sign
**** offline-verify-sign
**** serial-num
**** state

*** Sets Defaults Key Usage

See the documentation for further details.

--------------------------------------------------------------------
This file is part of EClet.

EClet is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
any later version.

EClet is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with EClet .  If not, see <http://www.gnu.org/licenses/>.

Local variables:
mode: outline
paragraph-separate: "[ 	]*$"
end:


src/cli/main.c
/* -*- mode: c; c-file-style: "gnu" -*-
 * Copyright (C) 2013 Cryptotronix, LLC.
 *
 * This file is part of EClet.
 *
 * EClet is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * EClet is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with EClet.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

/**
 * @file   main.c
 * @author Joshua Datko <jbd@cryptotronix.com>
 * @date   Sat May 24 20:47:39 2014
 *
 * @brief  Main entry point for EClet command line driver
 *
 *
 */

#include <argp.h>
#include <assert.h>
#include "cli_commands.h"
#include "config.h"
#include <string.h>
#include <libcrypti2c.h>

const char *argp_program_version = PACKAGE_VERSION;

const char *argp_program_bug_address = PACKAGE_BUGREPORT;

/* Program documentation. */
static char doc[] =
  "EClet is a program to interface to the Cryptotronix EClet which contains\n"
  "an Atmel ATECC108\n\n"
  "Currently implemented Commands:\n\n"
  "personalize   --  You should run this command first upon receiving your\n"
  "                  EClet.\n"
  "random        --  Retrieves 32 bytes of random data from the device.\n"
  "serial-num    --  Retrieves the device's serial number.\n"
  "get-config    --  Dumps the configuration zone\n"
  "get-otp       --  Dumps the OTP (one time programmable) zone\n"
  "state         --  Returns the device's state.\n"
  "                  Factory -- Random will produced a fixed 0xFFFF0000\n"
  "                  Initialized -- Configuration is locked, keys may be \n"
  "                                 written\n"
  "                  Personalized -- Keys are loaded.  Memory is locked\n"
  "gen-key       --  Generates a P256 Private key in the specified key slot\n"
  "                  Returns the Public Key (x,y) with the leading uncompressed\n"
  "                  point format tag (0x04)\n"
  "sign          --  Performs an ECDSA signature using the NIST P-256 curve.\n"
  "                  Specify the file to signed with -f, which will be SHA-256\n"
  "                  hashed prior to signing. Specify the key with -k.\n"
  "                  Returns the signature (R,S)\n"
  "verify        --  Uses the device to verify the signature.\n"
  "                  Specify the public key with --public-key, you must include\n"
  "                    the 0x04 tag followed by xy\n"
  "                  Specify the signature with --signature\n"
  "                  Specify the file with -f, it will be hashed with SHA256\n"
  "offline-verify-sign\n"
  "              --  Same as verify except it does NOT use the device, but a \n"
  "                  software library.";


/* A description of the arguments we accept. */
static char args_doc[] = "command";

#define OPT_UPDATE_SEED 300
#define OPT_SIGNATURE 301
#define OPT_PUB_KEY 302

/* The options we understand. */
static struct argp_option options[] = {
  { 0, 0, 0, 0, "Global Options:", -1},
  {"verbose",  'v', 0,      0,  "Produce verbose output" },
  {"quiet",    'q', 0,      0,  "Don't produce any output" },
  {"silent",   's', 0,      OPTION_ALIAS },
  {"bus",      'b', "BUS",  0,  "I2C bus: defaults to /dev/i2c-1"},
  {"address",  'a', "ADDRESS",      0,  "i2c address for the device (in hex)"},
  {"file",     'f', "FILE",         0,  "Read from FILE vs. stdin"},
  { 0, 0, 0, 0, "Sign and Verify Operations:", 1},
  {"signature", OPT_SIGNATURE, "SIGNATURE", 0, "The signature to be verified"},
  {"public-key", OPT_PUB_KEY, "PUBLIC_KEY", 0,
   "The public key that produced the signature"},
  { 0, 0, 0, 0, "Random Command Options:", 2},
  {"update-seed", OPT_UPDATE_SEED, 0, 0,
     "Updates the random seed.  Only applicable to certain commands"},
  { 0, 0, 0, 0, "Key related command options:", 3},
  {"key-slot", 'k', "SLOT",      0,  "The internal key slot to use."},
  {"write", 'w', "WRITE",      0,
   "The 32 byte data to write to a slot (64 bytes of ASCII Hex)"},
  { 0, 0, 0, 0, "Check and Offline-Verify Mac Options:", 4},
  {"challenge", 'c', "CHALLENGE",      0,
   "The 32 byte challenge (64 bytes of ASCII Hex)"},
  {"challenge-response", 'r', "CHALLENGE_RESPONSE",      0,
   "The 32 byte challenge response (64 bytes of ASCII Hex)"},
  {"meta-data", 'm', "META",      0,
   "The 13 byte meta data associated with the mac (26 bytes of ASCII Hex)"},
  { 0 }
};


/* Parse a single option. */
static error_t
parse_opt (int key, char *arg, struct argp_state *state)
{
  /* Get the input argument from argp_parse, which we
     know is a pointer to our arguments structure. */
  struct arguments *arguments = state->input;
  int slot;
  long int address_arg;

  switch (key)
    {
    case 'a':

      /* TODO: Not working as expected */
      address_arg = strtol (arg,NULL,16);
      if (0 != address_arg)
        {
          arguments->address = address_arg;
          CI2C_LOG (DEBUG, "Using address %u", address_arg);
        }
      else
        CI2C_LOG (INFO, "Address not recognized, using default");
    case 'b':
      arguments->bus = arg;
      break;
    case 'q': case 's':
      arguments->silent = 1;
      break;
    case 'v':
      arguments->verbose = 1;
      ci2c_set_log_level (DEBUG);
      break;
    case 'f':
      arguments->input_file = arg;
      break;
    case OPT_UPDATE_SEED:
      arguments->update_seed = true;
      break;
    case 'k':
      slot = atoi (arg);
      if (slot < 0 || slot > 15)
        argp_usage (state);

      arguments->key_slot = slot;
      break;
    case 'c':
      if (!is_hex_arg (arg, 64))
        {
          fprintf (stderr, "%s\n", "Invalid Challenge.");
          argp_usage (state);
        }
      else
        arguments->challenge = arg;
      break;
    case OPT_SIGNATURE:
      if (!is_hex_arg (arg, 128))
        {
          fprintf (stderr, "%s\n", "Invalid P256 Signature.");
          argp_usage (state);
        }
      else
      arguments->signature = arg;
      break;
    case OPT_PUB_KEY:
      if (!is_hex_arg (arg, 130))
        {
          fprintf (stderr, "%s\n", "Invalid P256 Public Key.");
          argp_usage (state);
        }
      else
        arguments->pub_key = arg;
      break;
    case 'w':
      if (!is_hex_arg (arg, 64))
        {
          fprintf (stderr, "%s\n", "Invalid Data.");
          argp_usage (state);
        }
      else
        arguments->write_data = arg;
      break;
    case 'r':
      if (!is_hex_arg (arg, 64))
        {
          fprintf (stderr, "%s\n", "Invalid Challenge Response.");
          argp_usage (state);
        }
      else
        arguments->challenge_rsp = arg;
      break;
    case 'm':
      if (!is_hex_arg (arg, 26))
        {
          fprintf (stderr, "%s\n", "Invalid Meta Data.");
          argp_usage (state);
        }
      else
        arguments->meta = arg;
      break;
    case ARGP_KEY_ARG:
      if (state->arg_num > NUM_ARGS)
        /* Too many arguments. */
        argp_usage (state);

      arguments->args[state->arg_num] = arg;

      break;

    case ARGP_KEY_END:
      if (state->arg_num < NUM_ARGS)
        /* Not enough arguments. */
        argp_usage (state);
      break;

    default:
      return ARGP_ERR_UNKNOWN;
    }
  return 0;
}

/* Our argp parser. */
static struct argp argp = { options, parse_opt, args_doc, doc };

int main (int argc, char **argv)
{
  struct arguments arguments;

  /* Sets arguments defaults and the command list */
  init_cli (&arguments);

  /* Parse our arguments; every option seen by parse_opt will
     be reflected in arguments. */
  argp_parse (&argp, argc, argv, 0, 0, &arguments);

  exit (dispatch (arguments.args[0], &arguments));

}
